src/Collapse/include/gudhi/Flag_complex_sparse_matrix.h

```diff
@@ -0,0 +1,424 @@
+            auto neighbours_c = closed_neighbours_row_index(rw_c);
+            // If neighbours_c contains the common neighbours.
+            if (std::includes(neighbours_c.begin(), neighbours_c.end(), common_neighbours.begin(),
+                              common_neighbours.end()))
+              return true;
+          }
+        }
+    }
+    return false;
+  }
+
+  std::set<Row_index> three_clique_indices(Row_index crit) {
+    std::set<Row_index> edge_indices;
+
+    Edge edge = std::get<0>(f_edge_vector_[crit]);
+    Vertex_handle u = std::get<0>(edge);
+    Vertex_handle v = std::get<1>(edge);
+
+#ifdef DEBUG_TRACES
+    std::cout << "The  current critical edge to re-check criticality with filt value is : f {" << u << "," << v
+              << "} = " << std::get<1>(f_edge_vector_[crit]) << std::endl;
+#endif  // DEBUG_TRACES
+    auto rw_u = vertex_to_row_[u];
+    auto rw_v = vertex_to_row_[v];
+    auto rw_critical_edge = std::make_pair(rw_u, rw_v);
+
+    Row_indices_vector common_neighbours = closed_common_neighbours_row_index(rw_critical_edge);
+
+    if (common_neighbours.size() > 2) {
+      for (auto rw_c : common_neighbours) {
+        if (rw_c != rw_u and rw_c != rw_v) {
+          auto e_with_new_nbhr_v = std::minmax(u, row_to_vertex_[rw_c]);
+          auto e_with_new_nbhr_u = std::minmax(v, row_to_vertex_[rw_c]);
+          edge_indices.emplace(edge_to_index_map_[e_with_new_nbhr_v]);
+          edge_indices.emplace(edge_to_index_map_[e_with_new_nbhr_u]);
+        }
+      }
+    }
+    return edge_indices;
+  }
+
+  // Detect and set all indices that are becoming critical
+  template<typename FilteredEdgeInsertion>
+  void set_edge_critical(Row_index indx, Filtration_value filt, FilteredEdgeInsertion filtered_edge_insert) {
+#ifdef DEBUG_TRACES
+    std::cout << "The curent index  with filtration value " << indx << ", " << filt << " is primary critical" <<
+    std::endl;
+#endif  // DEBUG_TRACES
+    std::set<Row_index> effected_indices = three_clique_indices(indx);
+    if (effected_indices.size() > 0) {
+      for (auto idx = indx - 1; idx > 0; idx--) {
+        Edge edge = std::get<0>(f_edge_vector_[idx]);
+        Vertex_handle u = std::get<0>(edge);
+        Vertex_handle v = std::get<1>(edge);
+        // If idx is not critical so it should be processed, otherwise it stays in the graph
+        if (!critical_edge_indicator_[idx]) {
+          // If idx is affected
+          if (effected_indices.find(idx) != effected_indices.end()) {
+            if (!check_edge_domination(edge)) {
+#ifdef DEBUG_TRACES
+              std::cout << "The curent index became critical " << idx  << std::endl;
+#endif  // DEBUG_TRACES
+              critical_edge_indicator_[idx] = true;
+              filtered_edge_insert({u, v}, filt);
+              std::set<Row_index> inner_effected_indcs = three_clique_indices(idx);
+              for (auto inr_idx = inner_effected_indcs.rbegin(); inr_idx != inner_effected_indcs.rend(); inr_idx++) {
```

Comment:
Any particular reason to do this backwards? If not, a for range loop would make sense.

Reply:
Range loop on 9ae3c860